File: apps/api/src/integrations/_mocks/store.ts
// Shared in-memory state for the mock layer. Tests reset it between runs.
// Mocks read from + write to this store; route handlers project from it.

export type MockInvoice = {
  id: string;
  customer_id: string;
  customer_name: string;
  doc_number: string;
  amount_cents: number;
  issued_at: string; // ISO
  due_at: string;
  paid_at: string | null;
  status: 'draft' | 'sent' | 'paid' | 'voided';
  source: 'qbo' | 'harvest';
};

export type MockPayment = {
  id: string;
  customer_name_raw: string;
  amount_cents: number;
  received_at: string;
  applied_to_invoice_id: string | null;
};

export type MockContract = {
  id: string; // signature_request_id
  title: string;
  recipient: string;
  status: 'sent' | 'viewed' | 'signed' | 'declined' | 'expired';
  sent_at: string;
  signed_at: string | null;
};

export type MockTimeEntry = {
  id: string;
  user_id: string;
  project_id: string;
  date: string; // YYYY-MM-DD
  hours: number;
  notes: string | null;
};

export type MockGmailMessage = {
  id: string;
  thread_id: string;
  from: string;
  to: string[];
  cc: string[];
  subject: string;
  body: string;
  sent_at: string;
  label_ids: string[];
};

// Simulate next-write-fails for testing failure paths.
// Tests call: failures.next('qbo:applyPayment', { status: 401 })
export type FailureSpec = { status: number; body?: unknown };

class FailureRegistry {
  private q = new Map<string, FailureSpec[]>();
  next(key: string, spec: FailureSpec): void {
    const list = this.q.get(key) ?? [];
    list.push(spec);
    this.q.set(key, list);
  }
  consume(key: string): FailureSpec | null {
    const list = this.q.get(key);
    if (!list || list.length === 0) return null;
    const next = list.shift()!;
    if (list.length === 0) this.q.delete(key);
    return next;
  }
  reset(): void {
    this.q.clear();
  }
}

export type MockStore = {
  invoices: Map<string, MockInvoice>;
  payments: Map<string, MockPayment>;
  contracts: Map<string, MockContract>;
  time_entries: Map<string, MockTimeEntry>;
  gmail_messages: Map<string, MockGmailMessage>;
  // sent emails the system claims to have delivered
  sent_emails: Array<{ to: string; cc: string[]; subject: string; body: string; thread_id: string | null; at: string }>;
  // contract reminders sent
  contract_reminders: Array<{ signature_request_id: string; at: string }>;
  // airtable records and field-mapped projects
  airtable_records: Map<string, Record<string, unknown>>;
  // harvest projects created
  harvest_projects: Map<string, { id: string; name: string; client_id: string; budget_hours: number }>;
  // drive folders created (no body, just a name)
  drive_folders: Array<{ path: string; created_at: string }>;
  failures: FailureRegistry;
};

const _store: MockStore = {
  invoices: new Map(),
  payments: new Map(),
  contracts: new Map(),
  time_entries: new Map(),
  gmail_messages: new Map(),
  sent_emails: [],
  contract_reminders: [],
  airtable_records: new Map(),
  harvest_projects: new Map(),
  drive_folders: [],
  failures: new FailureRegistry(),
};

export function getMockStore(): MockStore {
  return _store;
}

export function resetMockStore(): void {
  _store.invoices.clear();
  _store.payments.clear();
  _store.contracts.clear();
  _store.time_entries.clear();
  _store.gmail_messages.clear();
  _store.sent_emails.length = 0;
  _store.contract_reminders.length = 0;
  _store.airtable_records.clear();
  _store.harvest_projects.clear();
  _store.drive_folders.length = 0;
  _store.failures.reset();
}

File: apps/api_py/src/linkbook/integrations/mocks/store.py
"""Mock store — singleton dict-of-dicts shared across mock handlers.

Mirrors apps/api/src/integrations/_mocks/store.ts.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass
class MockInvoice:
    id: str
    customer_id: str
    customer_name: str
    doc_number: str
    amount_cents: int
    issued_at: str
    due_at: str
    paid_at: str | None
    status: str  # draft|sent|paid|voided
    source: str  # qbo|harvest


@dataclass
class MockPayment:
    id: str
    customer_name_raw: str
    amount_cents: int
    received_at: str
    applied_to_invoice_id: str | None


@dataclass
class MockContract:
    id: str
    title: str
    recipient: str
    status: str  # sent|viewed|signed|declined|expired
    sent_at: str
    signed_at: str | None


@dataclass
class MockTimeEntry:
    id: str
    user_id: str
    project_id: str
    date: str
    hours: float
    notes: str | None


@dataclass
class MockGmailMessage:
    id: str
    thread_id: str
    from_: str
    to: list[str]
    cc: list[str]
    subject: str
    body: str
    sent_at: str
    label_ids: list[str]


@dataclass
class FailureSpec:
    status: int
    body: Any | None = None


class FailureRegistry:
    def __init__(self) -> None:
        self._q: dict[str, list[FailureSpec]] = {}

    def queue_next(self, key: str, spec: FailureSpec) -> None:
        self._q.setdefault(key, []).append(spec)

    def consume(self, key: str) -> FailureSpec | None:
        lst = self._q.get(key)
        if not lst:
            return None
        spec = lst.pop(0)
        if not lst:
            self._q.pop(key, None)
        return spec

    def reset(self) -> None:
        self._q.clear()


@dataclass
class MockStore:
    invoices: dict[str, MockInvoice] = field(default_factory=dict)
    payments: dict[str, MockPayment] = field(default_factory=dict)
    contracts: dict[str, MockContract] = field(default_factory=dict)
    time_entries: dict[str, MockTimeEntry] = field(default_factory=dict)
    gmail_messages: dict[str, MockGmailMessage] = field(default_factory=dict)
    sent_emails: list[dict[str, Any]] = field(default_factory=list)
    contract_reminders: list[dict[str, Any]] = field(default_factory=list)
    airtable_records: dict[str, dict[str, Any]] = field(default_factory=dict)
    harvest_projects: dict[str, dict[str, Any]] = field(default_factory=dict)
    harvest_clients: dict[str, dict[str, Any]] = field(default_factory=dict)
    drive_folders: list[dict[str, Any]] = field(default_factory=list)
    failures: FailureRegistry = field(default_factory=FailureRegistry)


_STORE = MockStore()


def get_mock_store() -> MockStore:
    return _STORE


def reset_mock_store() -> None:
    s = _STORE
    s.invoices.clear()
    s.payments.clear()
    s.contracts.clear()
    s.time_entries.clear()
    s.gmail_messages.clear()
    s.sent_emails.clear()
    s.contract_reminders.clear()
    s.airtable_records.clear()
    s.harvest_projects.clear()
    s.harvest_clients.clear()
    s.drive_folders.clear()
    s.failures.reset()
